src/main/java/org/dsa/iot/alarm/Subscriptions.java

```diff
@@ -0,0 +1,199 @@
+                                       Handler<SubscriptionValue> handler) {
+        SubscriptionHandler facade = subscriptions.get(path);
+        if (facade == null) {
+            facade = new SubscriptionHandler(handler);
+            subscriptions.put(path, facade);
+            AlarmUtil.enqueue(new Runnable() {
+                @Override
+                public void run() {
+                    subscribeActual(path);
+                }
+            });
+        } else {
+            facade.add(handler);
+        }
+    }
+
+    /**
+     * Asynchronously called by subscribe to perform the actual subscription with the
+     * requester.
+     */
+    private void subscribeActual(String path) {
+        try {
+            SubscriptionHandler facade = null;
+            synchronized (this) {
+                facade = subscriptions.get(path);
+            }
+            if (facade == null) {
+                return;
+            }
+            Requester requester = getRequester();
+            if (requester != null) {
+                requester.subscribe(path, facade);
+            }
+        } catch (Exception x) {
+            AlarmUtil.logError(path, x);
+        }
+    }
+
+    public synchronized void unsubscribe(final String path,
+                                         Handler<SubscriptionValue> handler) {
+        SubscriptionHandler facade = subscriptions.get(path);
+        facade.remove(handler);
+        if (facade.size() == 0) {
+            subscriptions.remove(path);
+            if (service.getLinkHandler().getRequesterLink().isConnected()) {
+                AlarmUtil.enqueue(new Runnable() {
+                    @Override
+                    public void run() {
+                        unsubscribeActual(path);
+                    }
+                });
+            }
+        }
+    }
+
+    /**
+     * Asynchronously called by unsubscribe to perform the actual unsubscription with the
+     * requester.
+     */
+    private void unsubscribeActual(String path) {
+        try {
+            Requester requester = getRequester();
+            if (requester != null) {
+                requester.unsubscribe(path, null);
+            }
+        } catch (Exception x) {
+            AlarmUtil.logError(path, x);
+        }
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Inner Classes
+    ///////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Acts as a handler for multiple subscriptions to the same path.  However, most
+     * paths will have only one subscriber, so this is optimized for that.
+     */
+    private static class SubscriptionHandler implements Handler<SubscriptionValue> {
+
+        private Handler<SubscriptionValue> initialHandler;
+        private Set<Handler<SubscriptionValue>> handlers;
+
+        public SubscriptionHandler(Handler<SubscriptionValue> initialHandler) {
+            this.initialHandler = initialHandler;
```

Comment:
I have just read this `most paths will have only one subscriber, so this is optimized for that.` so I understand the reasoning behind it. 

However, I think it'd simplify the code quite a bit if you used a LinkedList instead of this, and there wouldn't be much performance loss, and yet you'd save on clarity, and also on performance compared to say, with an ArrayList.